### PR TITLE
August dev

### PR DESCRIPTION
**Pull request by Tsugua977**
-Added code where it gets tower and sprite positions.
-Tower can figure out which enemy is closest to it.
-Made tower ready for shooting.

### CODE REVIEW

File: entities/tower.py [hunk 1]
```diff
@@ -1,22 +1,106 @@
+import arcade
 from arcade import SpriteList
 from entities.entity import entity
 
-
 class tower(entity):
 
+    all_enemies:SpriteList = None
```
Comment (codingthunder) [inline entities/tower.py]: Put a space between all_enemies : SpriteList

File: entities/tower.py [hunk 2]
```diff
@@ -1,22 +1,106 @@
...
+        #Value of the range a tower can shoot.
+        self.range = 50
+
 
     def on_update(self, delta_time: float = 1 / 60):
         super().on_update(delta_time)
         # implement logic
+        #Tracks rate of fire.
+
+        def towerShoot():
```
Comment (codingthunder) [inline entities/tower.py:35]: Fix indentation.

File: entities/tower.py [hunk 3]
```diff
@@ -1,22 +1,106 @@
...
+
+                self.time_since_last_firing = 0
+                #print("Bang")
+                #Bullet spawn positon to tower goes here. 
+
+        #If the resources and tower level are correct, then the tower is upgraded. If the player does not have
+        #the correct number of resources, then every value stays the same and the user is told they don't have
+        #enough resources.
+        def upgrade(resources, damage, tower_level):
+            if resources >= 30 and tower_level == 3:
```
Comment (codingthunder) [inline entities/tower.py:48]: We don't want to hardcode level values. It should be procedural or data-driven.

File: entities/tower.py [hunk 4]
```diff
@@ -1,22 +1,106 @@
...
+        #Testing code for upgrade() to make sure it works.
+        # upgrade_results = upgrade(self.resources_value, self.damage_multiplier, self.current_tower_lv)
+        # resources_value = upgrade_results[0]
+        # damage_multiplier = upgrade_results[1]
+        # current_tower_lv = upgrade_results[2]
+        # print(f"Rsources Left: {resources_value}\nCurrent Damage: {damage_multiplier}\nTower Level: {current_tower_lv}")
+        
+        # 
+        def check_resources(resources):
+            if resources == 0:
```
Comment (codingthunder) [inline entities/tower.py]: See above comment as well. The resources checks are redundant. You should be able to determine if you have sufficient resources in one line of code. And if not, then just message "insufficient resources."

File: entities/tower.py [hunk 5]
```diff
@@ -1,22 +1,106 @@
...
+        # Test code for checking resources.
+        # resource_check = check_resources(self.resources_value)
+        # print(resource_check)
+
+        # spriteDistance = arcade.get_distance_between_sprites(
+        #     tower = arcade.sprite.tower,
+        #     enemies = arcade.sprite.enemy
+        # )
+
+        targetedEnemy = arcade.get_closest_sprite(
```
Comment (codingthunder) [inline entities/tower.py]: Check your indenting. What function is this supposed to be in?

File: entities/tower.py [hunk 6]
```diff
@@ -1,22 +1,106 @@
...
+        def check_resources(resources):
+            if resources == 0:
+                print("No resources")
+                upgrade_results = "None"
+                return upgrade_results
+            else:
+                upgrade_results = upgrade(self.resources_value, self.damage_multiplier, self.current_tower_lv)
+                return upgrade_results
+
+        # Test code for checking resources.
```
Comment (codingthunder) [inline entities/tower.py]: The commented out code, if unnecessary, should be deleted. It disrupts the visual flow of code and makes it harder to understand.

File: gameArchitecture/dataManager.py [hunk 7]
```diff
@@ -26,7 +26,7 @@ def level_dummy_data() -> level_data:
         word_data = ["Coconuts", "Make", "Me", "Giggle"]
         for x in range(10):
             route.append((9 * 50, (9 - x) * 50))
-            towers.append((350, x * 50))
+            towers.append((350, 150))
```
Comment (codingthunder) [inline gameArchitecture/dataManager.py:29]: Not super pertinent, but when we start laying out the levels, we'll need to assign this based on data.

**Review by codingthunder — CHANGES_REQUESTED** [review]
See my code comments.

File: entities/tower.py (hunk 3 above)
Comment (Tsugua977) [reply entities/tower.py:48]: I put this in as a temporary solution to test the code until we could get data driven code in there.

File: gameArchitecture/dataManager.py (hunk 7 above)
Comment (Tsugua977) [reply gameArchitecture/dataManager.py:29]: Alright, sounds good! I just did that for testing purposes.

**Comment by Tsugua977** [comment]
Fixed all comments that needed to be fixed. If there were any I did not fix, I commented on why I didn't. If anything else needs to be changed, let me know.

**Pull request merged**